modules/fedimint-lnv2-server/src/lib.rs

```diff
@@ -0,0 +1,668 @@
+                    .await
+                    .unwrap_or(0);
+
+                ensure!(current_vote < vote, "Block count vote is redundant");
+
+                Ok(())
+            }
+            LightningConsensusItem::UnixTimeVote(vote) => {
+                let current_vote = dbtx
+                    .insert_entry(&UnixTimeVoteKey(peer), &vote)
+                    .await
+                    .unwrap_or(0);
+
+                ensure!(current_vote < vote, "Unix time vote is redundant");
+
+                Ok(())
+            }
+        }
+    }
+
+    async fn process_input<'a, 'b, 'c>(
+        &'a self,
+        dbtx: &mut DatabaseTransaction<'c>,
+        input: &'b LightningInput,
+    ) -> Result<InputMeta, LightningInputError> {
+        let (pub_key, amount) = match &input.witness {
+            Witness::Outgoing(contract_id, outgoing_witness) => {
+                let contract = dbtx
+                    .remove_entry(&OutgoingContractKey(*contract_id))
+                    .await
+                    .ok_or(LightningInputError::UnknownContract)?;
+
+                match outgoing_witness {
+                    OutgoingWitness::Claim(preimage) => {
+                        if contract.expiration <= self.consensus_block_count(dbtx).await {
+                            return Err(LightningInputError::Expired);
+                        }
+
+                        if !contract.verify_preimage(preimage) {
+                            return Err(LightningInputError::InvalidPreimage);
+                        }
+
+                        dbtx.insert_new_entry(&PreimageKey(*contract_id), preimage)
+                            .await;
+
+                        (contract.claim_pk, contract.amount)
+                    }
+                    OutgoingWitness::Refund => {
+                        if contract.expiration > self.consensus_block_count(dbtx).await {
+                            return Err(LightningInputError::NotExpired);
+                        }
+
+                        (contract.refund_pk, contract.amount)
+                    }
+                    OutgoingWitness::Cancel(forfeit_signature) => {
+                        if !contract.verify_forfeit_signature(forfeit_signature) {
+                            return Err(LightningInputError::InvalidForfeitSignature);
+                        }
+
+                        (contract.refund_pk, contract.amount)
+                    }
+                }
+            }
+            Witness::Incoming(contract_id, agg_decryption_key) => {
+                let contract = dbtx
+                    .remove_entry(&IncomingContractKey(*contract_id))
+                    .await
+                    .ok_or(LightningInputError::UnknownContract)?;
+
+                if !contract
+                    .verify_agg_decryption_key(&self.cfg.consensus.tpe_agg_pk, agg_decryption_key)
+                {
+                    return Err(LightningInputError::InvalidDecryptionKey);
+                }
```

Comment:
Do we need to check contract expiry here too? I only see `consensus_unix_time` being checked in `process_output`, just trying to understand.